packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts

```diff
@@ -180,6 +203,17 @@ ruleTester.run('prefer-optional-chain', rule, {
     'foo && foo[bar as string] && foo[bar as string].baz;',
     'foo && foo[1 + 2] && foo[1 + 2].baz;',
     'foo && foo[typeof bar] && foo[typeof bar].baz;',
+    '!foo || !foo[bar as string] || !foo[bar as string].baz;',
+    '!foo || !foo[1 + 2] || !foo[1 + 2].baz;',
+    '!foo || !foo[typeof bar] || !foo[typeof bar].baz;',
+    // currently do not handle 'this' as the first part of a chain
+    'this && this.foo;',
+    '!this || !this.foo;',
+    // currently do not handle mixed TSNonNullExpression in properties
+    '!entity.__helper!.__initialized || options.refresh;',
```

Comment:
I think it makes sense that we would explicitly ignore `!`s, as they're often a sign of folks doing odd stuff in the type system.

```suggestion
    // intentionally do not handle mixed TSNonNullExpression in properties
    '!entity.__helper!.__initialized || options.refresh;',
```